rust/hyperlane-base/src/contract_sync/cursors/sequence_aware/forward.rs

```diff
@@ -0,0 +1,1089 @@
+        expected_sequence_range: &RangeInclusive<u32>,
+    ) {
+        warn!(
+            all_log_sequences=?all_log_sequences.iter().sorted().collect::<Vec<_>>(),
+            expected_sequences=?expected_sequences.iter().sorted().collect::<Vec<_>>(),
+            ?expected_sequence_range,
+            missing_expected_sequences=?expected_sequences.difference(all_log_sequences).sorted().collect::<Vec<_>>(),
+            unexpected_sequences=?all_log_sequences.difference(expected_sequences).sorted().collect::<Vec<_>>(),
+            ?logs,
+            current_indexing_snapshot=?self.current_indexing_snapshot,
+            last_indexed_snapshot=?self.last_indexed_snapshot,
+            target_snapshot=?self.target_snapshot,
+            "Log sequences don't exactly match the expected sequence range, rewinding to last indexed snapshot",
+        );
+        // If there are any missing sequences, rewind to index immediately after the last snapshot.
+        self.rewind();
+    }
+
+    // Rewinds the cursor to target immediately after the last indexed snapshot.
+    fn rewind(&mut self) {
+        self.current_indexing_snapshot = self.last_indexed_snapshot.next_target();
+    }
+}
+
+#[async_trait]
+impl<T: Sequenced + Debug> ContractSyncCursor<T> for ForwardSequenceAwareSyncCursor<T> {
+    async fn next_action(&mut self) -> Result<(CursorAction, Duration)> {
+        // TODO: Fix ETA calculation
+        let eta = Duration::from_secs(0);
+        if let Some(range) = self.get_next_range().await? {
+            Ok((CursorAction::Query(range), eta))
+        } else {
+            // TODO: Define the sleep time from interval flag
+            Ok((CursorAction::Sleep(Duration::from_secs(5)), eta))
+        }
+    }
+
+    // TODO: revisit to establish a better heuristic for cursor / indexing health
+    fn latest_block(&self) -> u32 {
+        self.current_indexing_snapshot.at_block
+    }
+
+    /// Updates the cursor with the logs that were found in the range.
+    ///
+    /// Inconsistencies in the logs are not considered errors, instead they're handled by rewinding the cursor
+    /// to retry ranges.
+    ///
+    /// ## logs
+    /// The logs to ingest. If any logs are duplicated or their sequence is lower than the current indexing snapshot,
+    /// they are filtered out.
+    ///
+    /// If Sequence mode:
+    /// - The sequences of the logs must exactly match the range.
+    /// - If there are any gaps, the cursor rewinds and the range will be retried.
+    ///
+    /// If Block mode:
+    /// - Empty logs are allowed, but no gaps are allowed. The logs must build upon the last indexed snapshot.
+    /// - If there are any gaps, the cursor rewinds to the last indexed snapshot, and ranges will be retried.
+    /// - If the target block is reached and the target sequence hasn't been reached, the cursor rewinds to the last indexed snapshot.
+    async fn update(&mut self, logs: Vec<(T, LogMeta)>, range: RangeInclusive<u32>) -> Result<()> {
```

Comment:
mm this is a good point

I think my initial goal was to have internal state only relate to things in service of making sure that we're building upon the last indexed sequence, and to have `update` avoid making assumptions about how the cursor is used and to just ensure that the logs passed in build upon the last indexed sequence, but I think the delineation isn't super clear. I'll come back to this

Reply:
Ideally the cursor implemented the `Iterator` trait but that's not really possible since it has internal state to mutate

Reply:
After mulling it over I think having the range explicitly passed in is probably helpful. Curious what you think

The cursor update is really dependent on the logs + range, not just the logs. I think it also could reduce misuse of cursors in the future - by putting a range in the function signature it makes it clear that the cursor updates rely heavily on the range. I think part of the messiness of the prior implementations was that it wasn't super clear what was happening inside a cursor or how a cursor should be used, I think if we put the range as internal state we now implicitly closely tie the cursor's `update` fn to the `get_action` fn

Reply:
Hmm I think there's potential for misuse if the range is required too, since it's not entirely clear that the range to pass in should be the one from `get_action`. I'm ok either way, just wanted to point this out